src/EditorFeatures/Core/FindUsages/FindUsagesHelpers.cs

```diff
@@ -47,95 +47,111 @@ public static string GetDisplayName(ISymbol symbol)
+            var symbolAndProjectId = new SymbolAndProjectId(symbol, project.Id);
+            return await FindSourceImplementationsAsync(
+                symbolAndProjectId, project.Solution, cancellationToken).ConfigureAwait(false);
         }
 
-        private static async Task<(ISymbol symbol, Project project, ImmutableArray<ISymbol> implementations, string message)?> FindImplementationsAsync(
-            ISymbol symbol, Project project, CancellationToken cancellationToken)
+        private static async Task<(Solution solution, SymbolAndProjectId symbolAndProjectId, ImmutableArray<SymbolAndProjectId> implementations, string message)?> FindSourceImplementationsAsync(
+            SymbolAndProjectId symbolAndProjectId, Solution solution, CancellationToken cancellationToken)
         {
-            var implementations = await FindImplementationsWorkerAsync(
-                symbol, project, cancellationToken).ConfigureAwait(false);
+            var builder = new HashSet<SymbolAndProjectId>(SymbolAndProjectIdComparer.SymbolEquivalenceInstance);
 
-            var filteredSymbols = implementations.WhereAsArray(
-                s => s.Locations.Any(l => l.IsInSource));
+            // Find the direct implementations first.
+            builder.AddRange(await FindSourceImplementationsWorkerAsync(
+                symbolAndProjectId, solution, cancellationToken).ConfigureAwait(false));
 
-            return filteredSymbols.Length == 0
-                ? (symbol, project, filteredSymbols, EditorFeaturesResources.The_symbol_has_no_implementations)
-                : (symbol, project, filteredSymbols, null);
+            // If we're in a linked file, try to find all the symbols this links to, and find all the implementations of
+            // each of those linked symbols. De-dupe the results so the user only gets unique results.
+            var linkedSymbols = await SymbolFinder.FindLinkedSymbolsAsync(
+                symbolAndProjectId.Symbol, solution, cancellationToken).ConfigureAwait(false);
+
+            foreach (var linkedSymbol in linkedSymbols)
+            {
+                builder.AddRange(await FindSourceImplementationsWorkerAsync(
+                    linkedSymbol, solution, cancellationToken).ConfigureAwait((bool)false));
+            }
+
+            var result = builder.ToImmutableArray();
+
+            return result.Length == 0
+                ? (solution, symbolAndProjectId, result, EditorFeaturesResources.The_symbol_has_no_implementations)
+                : (solution, symbolAndProjectId, result, null);
         }
 
-        private static async Task<ImmutableArray<ISymbol>> FindImplementationsWorkerAsync(
-            ISymbol symbol, Project project, CancellationToken cancellationToken)
+        private static async Task<ImmutableArray<SymbolAndProjectId>> FindSourceImplementationsWorkerAsync(
+            SymbolAndProjectId symbolAndProjectId, Solution solution, CancellationToken cancellationToken)
```

Comment:
💡 `solution` first for consistency

Reply:
Yup.  Happy to do that (in followup)! :)